sysutils/webdav-backup/src/opnsense/mvc/app/library/OPNsense/Backup/Webdav.php

```diff
@@ -0,0 +1,368 @@
+                return false;
+            }
+            // If error assume dir doesn't exist. Create parent folder
+            if ($this->create_directory($url, $username, $password, $parent_path) === false) {
+                throw new \Exception();
+            }
+        }
+        // if path exists ok
+        if (in_array("$backupdir", $directories)) {
+            return;
+        }
+
+        $this->curl_request(
+            $url . "/{$backupdir}",
+            $username,
+            $password,
+            'MKCOL',
+            'cannot execute MKCOL'
+        );
+    }
+
+    /**
+     * @param string $url remote location
+     * @param string $username remote user
+     * @param string $password password to use
+     * @param string $method http method, PUT, GET, ...
+     * @param string $error_message message to log on failure
+     * @param null|string $postdata http body
+     * @param array $headers HTTP headers
+     * @return array response status
+     * @throws \Exception when request fails
+     */
+    public function curl_request(
+        $url,
+        $username,
+        $password,
+        $method,
+        $error_message,
+        $postdata = null,
+        $headers = array("User-Agent: OPNsense Firewall")
+    ) {
+        $curl = curl_init();
+        curl_setopt_array($curl, array(
+            CURLOPT_URL => $url,
+            CURLOPT_CUSTOMREQUEST => $method, // Create a file in WebDAV is PUT
+            CURLOPT_RETURNTRANSFER => true, // Do not output the data to STDOUT
+            CURLOPT_VERBOSE => 0,           // same here
+            CURLOPT_MAXREDIRS => 0,         // no redirects
+            CURLOPT_TIMEOUT => 60,          // maximum time: 1 min
+            CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
+            CURLOPT_USERPWD => $username . ":" . $password,
+            CURLOPT_HTTPHEADER => $headers
+        ));
+        if ($postdata != null) {
+            curl_setopt($curl, CURLOPT_POSTFIELDS, $postdata);
+        }
+        $response = curl_exec($curl);
+        $err = curl_error($curl);
+        $info = curl_getinfo($curl);
+        if (!($info['http_code'] == 200 || $info['http_code'] == 207 || $info['http_code'] == 201) || $err) {
+            syslog(LOG_ERR, $error_message);
+            syslog(LOG_ERR, json_encode($info));
+            throw new \Exception();
+        }
+        curl_close($curl);
+        return array('response' => $response, 'info' => $info);
+    }
+
+    /**
+     * @param $url string URL to call
+     * @param $username string username
+     * @param $password string password
+     * @param $method string HTTP verb
+     * @param $error_message string error message to forward to the http calling method
+     * @param null $postdata post data if any (can be null)
+     * @return array|\SimpleXMLElement|null
+     * @throws \Exception
+     */
+    public function ocs_request($url, $username, $password, $method, $error_message, $postdata = null)
```

Comment:
ocs is for nextcloud, you can remove the  OCS header and the prefix of this method as they are for nextcloud

Reply:
I deleted it